lib/sockfd_cache.c

```diff
@@ -98,12 +101,24 @@ static inline int get_free_slot(struct sockfd_cache_entry *entry)
 {
 	int idx = -1, i;
 
+	if (entry->nid.io_port && entry->channel_status == IO) {
+		for (i = 0; i < fds_count; i++) {
+			if (entry->fds_io[i].fd == -1 ||
+				!uatomic_set_true(&entry->fds_io[i].in_use))
+				continue;
+			idx = i;
+			goto out;
+		}
+	}
 	for (i = 0; i < fds_count; i++) {
-		if (!uatomic_set_true(&entry->fds[i].in_use))
+		if (entry->fds_nio[i].fd == -1 ||
+			!uatomic_set_true(&entry->fds_nio[i].in_use))
 			continue;
-		idx = i;
-		break;
+		/*let caller know if this is an IO or NonIO port*/
+		idx = (i + 1) << 20;
```

Comment:
It seems a little bit hacky. How about let `get_free_slot()` have a new parameter of `enum channel_status*` and notify the type of socket to caller via the value?